libminifi/include/utils/GeneralUtils.h

```diff
@@ -68,6 +68,21 @@ using void_t = void;
 using std::void_t;
 #endif /* < C++17 */
 
+namespace internal {
+
+struct safe_enable_shared_from_this_base : std::enable_shared_from_this<safe_enable_shared_from_this_base> {
```

Comment:
we have to have a single class template instance `std::enable_shared_from_this<T>` for some type `T`, so that when we inherit from it virtually multiple times, we get the same `std::weak_ptr` instance

for example in the following, if `U` depends on `T` we don't inherit from the same `enable_shared_from_this` class instance, but `U` also must be a base class of `T`, moreover it has to be polymorphic to `dynamic_cast` it to the derived cast (that is why it has a virtual destructor)
```
template<typename T>
struct safe_enable_shared_from_this : virtual std::enable_shared_from_this<U> {...}

struct A : safe_enable_shared_from_this<A> {...}
struct B: A, safe_enable_shared_from_this<B>{...}
```

Reply:
Okay thanks, that's clear now.  Can you add a comment somewhere explaining that we need this code because `std::enable_shared_from_this` does not play nice with multiple inheritance (or a link to an explanation, if there is a good one somewhere public and durable)?

Reply:
added a comment